tsdb/shard.go

```diff
@@ -702,56 +702,64 @@ func (s *Shard) validateSeriesAndFields(points []models.Point, tracker StatsTrac
 			continue
 		}
 
-		name := p.Name()
-		mf := engine.MeasurementFields(name)
-
-		// Check with the field validator.
-		if err := ValidateFields(mf, p, s.options.Config.SkipFieldSizeValidation); err != nil {
-			switch err := err.(type) {
-			case PartialWriteError:
-				if reason == "" {
-					reason = err.Reason
+		cont, err := func(p models.Point, iter models.FieldIterator) (cont bool, err error) {
+			name := p.Name()
+			mf := engine.MeasurementFields(name)
+			mf.mu.Lock()
+			defer mf.mu.Unlock()
+			// Check with the field validator.
+			if err := ValidateFields(mf, p, s.options.Config.SkipFieldSizeValidation); err != nil {
+				switch err := err.(type) {
+				case PartialWriteError:
+					if reason == "" {
+						reason = err.Reason
+					}
+					dropped += err.Dropped
+					atomic.AddInt64(&s.stats.WritePointsDropped, int64(err.Dropped))
+				default:
+					return false, err
 				}
-				dropped += err.Dropped
-				atomic.AddInt64(&s.stats.WritePointsDropped, int64(err.Dropped))
-			default:
-				return nil, nil, err
+				return true, nil
 			}
-			continue
-		}
 
-		points[j] = points[i]
-		j++
+			points[j] = points[i]
+			j++
 
-		// Create any fields that are missing.
-		iter.Reset()
-		for iter.Next() {
-			fieldKey := iter.FieldKey()
+			// Create any fields that are missing.
+			iter.Reset()
+			for iter.Next() {
+				fieldKey := iter.FieldKey()
 
-			// Skip fields named "time". They are illegal.
-			if bytes.Equal(fieldKey, timeBytes) {
-				continue
-			}
+				// Skip fields named "time". They are illegal.
+				if bytes.Equal(fieldKey, timeBytes) {
+					continue
+				}
 
-			if mf.FieldBytes(fieldKey) != nil {
-				continue
-			}
+				if mf.FieldBytes(fieldKey) != nil {
+					continue
+				}
 
-			dataType := dataTypeFromModelsFieldType(iter.Type())
-			if dataType == influxql.Unknown {
-				continue
-			}
+				dataType := dataTypeFromModelsFieldType(iter.Type())
+				if dataType == influxql.Unknown {
+					continue
+				}
 
-			fieldsToCreate = append(fieldsToCreate, &FieldCreate{
-				Measurement: name,
-				Field: &Field{
-					Name: string(fieldKey),
-					Type: dataType,
-				},
-			})
+				fieldsToCreate = append(fieldsToCreate, &FieldCreate{
+					Measurement: name,
+					Field: &Field{
+						Name: string(fieldKey),
+						Type: dataType,
+					},
+				})
```

Comment:
If a write has two points that both have a new field (`newField`), but the two points have different types for `newField`, they would both end up in `fieldsToCreate`, correct? Would that create issues later?

Reply:
The old code had that same pattern.  The first gets created, the second rejected.